### ISSUE TITLE
Overriding the default() operator.

### ISSUE THREAD

**Issue by orthoxerox** [dotnet/roslyn#71]
I know there's been a discussion on StackOverflow involving Eric Lippert on the potential problems with overriding the default() operator, but could it be implemented with additional restrictions?

1. Overridden default() operator can return only a static readonly instance of the class or its subclass that have to be immutable. If the class is mutable, then the immutable subclass cannot access any base() methods or write to any fields/properties. Uh, I guess this is a whole 'nother issue, declaring immutability and asserting it by the compiler.

2. If it's a struct, parameterless constructor must be equivalent to default().

3. If it's a class, the compiler initializes all references to default() instead of null (I know it's slower, since you can't simply XOR the pointer with itself, but it's just copying a single value). Assigning null to it is rewritten to assigning default(T) by the compiler, the same with checking if it's null.

This way the consumer of the new class won't notice anything except the lack of NRE's.

**Comment by mikedn** [comment]
What problem would this solve anyway? Seems like a good way to confuse the readers of code which makes of this, not to mention that it's difficult or even impossible to get it to work with generics.

**Comment by orthoxerox** [comment]
Transparently replacing nulls with null objects and avoiding NREs.

**Comment by mikedn** [comment]
And what does avoiding NREs solve? NREs are the result of 2 type of bugs:

1. Failure to initialize a variable
2. Failure to check for null in cases where a null is expected

Let's what can happen if you replace the null with some kind of 'default object':

For case 1 you'll end up accessing this default object and the result will likely be intended. Either the object will perform no operation and in turn that will cause other problems (hey, I saved this document I was working on but the file is nowhere to be seen) or it will throw an exception (hey, you're not supposed to do this).

In case 2 the situation is slightly better, it will work fine in some cases but not others. Consider:

```C#
if (stream == null)
    return; // no stream, do nothing
stream.Write(...);
Console.WriteLine("document written");
```

If you forget the null check and stream is Stream.Null rather than null the result is that you'll end up telling the user that the document has been written when nothing actually happened and it wasn't supposed to happen anyway (remember than in this case where the null is expected).

In short, such an approach simply transforms one bug into another bug. And, arguably, the new bug is worse than the old one because silent failures are more difficult to notice and debug.

**Comment by paulomorgado** [comment]
Unce upon a time some said it's not cool to have `null`s.

**Comment by sharwell** [comment]
@mikedn is correct. The `NullReferenceException` indicates a misuse of some reference in the application. Changing the meaning of `default(T)` for some `T` would not correct the misuse, but it certainly would make it more difficult to identify and understand many kinds of misuse.

**Comment by AdamSpeight2008** [comment]
Similar Topic: [Overridable Default Values For Null?](https://roslyn.codeplex.com/discussions/560569)

**Comment by Richiban** [comment]
I'm not in favour of replacing `null` with a null-object, but it would be useful to be able to override `default(T)`. Some types really do have a sensible default value, such as "" for string, [] for arrays etc.

This could have performance implications though: if an array of `T` is initialised with `var arr = new T[10000]` and `default(T)` requires actually constructing an object instead of using null or 0 etc then you've just constructed 10,000 objects in one statement. Probably not what you want to happen.

**Comment by gafter** [comment]
Closing, as I cannot imagine anyone on the extended design team championing this over other work. @MadsTorgersen please reopen if you disagree.
